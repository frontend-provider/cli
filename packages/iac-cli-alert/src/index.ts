import {
  IncomingWebhook,
  IncomingWebhookDefaultArguments,
} from '@slack/webhook';

if (!process.env.IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL) {
  console.error(
    'Missing the IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL environment variable',
  );
  process.exit(1);
}

if (!process.env.IAC_SMOKE_TESTS_OS) {
  console.error('Missing the IAC_SMOKE_TESTS_OS environment variable');
  process.exit(1);
}

const IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL =
  process.env.IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL;
const IAC_SMOKE_TESTS_OS = process.env.IAC_SMOKE_TESTS_OS;

const slackWebhook = new IncomingWebhook(IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL);

async function sendSlackAlert() {
  console.log('IaC smoke tests failed. Sending Slack alert...');
  const args: IncomingWebhookDefaultArguments = {
    username: 'IaC CLI Alerts',
    text: `An Infrastructure as Code Smoke Tests job failed. \n Operating system - ${IAC_SMOKE_TESTS_OS} \n<https://github.com/snyk/cli/actions?query=workflow%3A+%22Infrastructure+as+Code+Smoke+Tests%22|Infrastructure as Code Smoke Tests Results>`,
    icon_emoji: 'snyk-iac',
  };
  await slackWebhook.send(args);
  console.log('Slack alert sent.');
}

async function run() {
  try {
    await sendSlackAlert();
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
}

process.on('uncaughtException', (err) => {
  console.error(err);
  process.exit(1);
});

// Exec
run();
