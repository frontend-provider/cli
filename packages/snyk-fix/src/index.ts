import * as debugLib from 'debug';
import { loadPlugin } from './plugins/load-plugin';
import { FixHandlerResult } from './plugins/types';

import { EntityToFix } from './types';

const debug = debugLib('snyk-fix:main');

export async function fix(
  entities: EntityToFix[],
): Promise<FixHandlerResult[]> {
  debug(`Requested to fix ${entities.length} projects.`);
  const entitiesPerType = groupEntitiesPerType(entities);
  const allResults: FixHandlerResult[] = [];
  // TODO: pMap this?
  for (const type of Object.keys(entitiesPerType)) {
    const handler = loadPlugin(type);
    const results = await handler(entitiesPerType[type]);
    allResults.push(results);
  }

  return allResults;
}

function groupEntitiesPerType(
  entities: EntityToFix[],
): {
  [type: string]: EntityToFix[];
} {
  const entitiesPerType: {
    [type: string]: EntityToFix[];
  } = {};
  for (const entity of entities) {
    const type = entity.scanResult.identity.type;
    if (entitiesPerType[type]) {
      entitiesPerType[type].push(entity);
      continue;
    }
    entitiesPerType[type] = [entity];
  }
  return entitiesPerType;
}
