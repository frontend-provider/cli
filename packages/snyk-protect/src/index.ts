#!/usr/bin/env node

import protect from './lib';

async function main() {
  const projectPath = process.cwd();
  await protect(projectPath);
}

if (require.main === module) {
  main();
}
