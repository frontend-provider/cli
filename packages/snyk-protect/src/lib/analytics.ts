import { AnalyticsPayload, ProtectResult } from './types';
import { postJson } from './http';
import { getApiBaseUrl } from './snyk-api';
import * as os from 'os';
import { readFileSync, existsSync } from 'fs';
import * as path from 'path';

function getVersion(): string {
  return JSON.parse(
    readFileSync(path.resolve(__dirname, '../../package.json'), 'utf-8'),
  ).version;
}

function getAnalyticsData(result: ProtectResult): AnalyticsPayload {
  return {
    command: `@snyk/protect`,
    args: [],
    version: getVersion(),
    nodeVersion: process.version,
    metadata: {
      protectResult: result,
    },
  };
}

export async function sendAnalytics(result: ProtectResult) {
  if (!allowAnalytics()) {
    return;
  }
  try {
    const apiBaseUrl = getApiBaseUrl();
    const apiUrl = `${apiBaseUrl}/v1/analytics/cli`;
    const data = {
      data: getAnalyticsData(result),
    };
    await postJson(apiUrl, data);
  } catch (err) {
    // do nothing and log nothing
  }
}

function allowAnalytics(): boolean {
  try {
    const snykConfigFile = getSnykConfigFilePath();
    if (existsSync(snykConfigFile)) {
      const config = JSON.parse(readFileSync(snykConfigFile, 'utf-8'));
      if (
        config['disable-analytics'] === '1' ||
        config['disable-analytics'] === 1
      ) {
        return false;
      }
    }
    if (process.env.SNYK_DISABLE_ANALYTICS) {
      return false;
    }
  } catch (err) {
    // do nothing and log nothing
  }
  return true;
}

function getSnykConfigFilePath(): string {
  return (
    process.env.SNYK_CONFIG_FILE ||
    path.resolve(os.homedir(), '.config', 'configstore', 'snyk.json')
  );
}
