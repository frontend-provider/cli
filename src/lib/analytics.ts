import { v4 as uuidv4 } from 'uuid';
const snyk = require('../lib');
const config = require('./config');
const version = require('./version');
const request = require('./request');
const {
  getIntegrationName,
  getIntegrationVersion,
  getIntegrationEnvironment,
  getIntegrationEnvironmentVersion,
  getCommandVersion,
} = require('./analytics-sources');
const isCI = require('./is-ci').isCI;
const debug = require('debug')('snyk');
const os = require('os');
const osName = require('os-name');
const crypto = require('crypto');
const stripAnsi = require('strip-ansi');
import * as needle from 'needle';
const { MetricsCollector } = require('./metrics');

const metadata = {};
// analytics module is required at the beginning of the CLI run cycle
const startTime = Date.now();

/**
 *
 * @param data the data to merge into that data which has been staged thus far (with the {@link add} function)
 * and then sent to the backend.
 */
export function addDataAndSend(
  data,
): Promise<void | { res: needle.NeedleResponse; body: any }> {
  if (!data) {
    data = {};
  }

  // merge any new data with data we picked up along the way
  if (Array.isArray(data.args)) {
    // this is an overhang from the cli/args.js and we don't want it
    delete (data.args.slice(-1).pop() || {})._;
  }

  if (Object.keys(metadata).length) {
    data.metadata = metadata;
  }

  return postAnalytics(data);
}

export function allowAnalytics(): boolean {
  if (snyk.config.get('disable-analytics') || config.DISABLE_ANALYTICS) {
    return false;
  } else {
    return true;
  }
}

/**
 * Actually send the analytics to the backend. This can be used standalone to send only the data
 * given by the data parameter, or called from {@link addDataAndSend}.
 * @param data the analytics data to send to the backend.
 */
export function postAnalytics(
  data,
): Promise<void | { res: needle.NeedleResponse; body: any }> {
  // if the user opt'ed out of analytics, then let's bail out early
  // ths applies to all sending to protect user's privacy
  if (!allowAnalytics()) {
    debug('analytics disabled');
    return Promise.resolve();
  }

  const isStandalone = version.isStandaloneBuild();

  // get snyk version
  return version
    .getVersion()
    .then(async (version) => {
      data.version = version;
      data.os = osName(os.platform(), os.release());
      data.nodeVersion = process.version;
      data.standalone = isStandalone;
      data.integrationName = getIntegrationName(data.args);
      data.integrationVersion = getIntegrationVersion(data.args);
      data.integrationEnvironment = getIntegrationEnvironment(data.args);
      data.integrationEnvironmentVersion = getIntegrationEnvironmentVersion(
        data.args,
      );

      const seed = uuidv4();
      const shasum = crypto.createHash('sha1');
      data.id = shasum.update(seed).digest('hex');

      const headers = {};
      if (snyk.api) {
        headers['authorization'] = 'token ' + snyk.api;
      }

      data.ci = isCI();

      data.environment = {};
      if (!isStandalone) {
        data.environment.npmVersion = await getCommandVersion('npm');
      }

      data.durationMs = Date.now() - startTime;

      try {
        const networkTime = MetricsCollector.NETWORK_TIME.getTotal();
        const cpuTime = data.durationMs - networkTime;
        MetricsCollector.CPU_TIME.createInstance().setValue(cpuTime);
        data.metrics = MetricsCollector.getAllMetrics();
      } catch (err) {
        debug('Error with metrics', err);
      }

      const queryStringParams = {};
      if (data.org) {
        queryStringParams['org'] = data.org;
      }

      debug('analytics', JSON.stringify(data, null, '  '));

      const queryString =
        Object.keys(queryStringParams).length > 0
          ? queryStringParams
          : undefined;

      return request({
        body: {
          data: data,
        },
        qs: queryString,
        url: config.API + '/analytics/cli',
        json: true,
        method: 'post',
        headers: headers,
      });
    })
    .catch((error) => {
      debug('analytics', error); // this swallows the analytics error
    });
}

/**
 * Adds a key-value pair to the analytics data `metadata` field. This doesn't send the analytis, just stages it for
 * sending later (via the {@link addDataAndSend} function).
 * @param key
 * @param value
 */
export function add(key, value) {
  if (typeof value === 'string') {
    value = stripAnsi(value);
  }
  if (metadata[key]) {
    if (!Array.isArray(metadata[key])) {
      metadata[key] = [metadata[key]];
    }
    metadata[key].push(value);
  } else {
    metadata[key] = value;
  }
}
