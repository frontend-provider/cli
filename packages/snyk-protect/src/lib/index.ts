import * as fs from 'fs';
import * as path from 'path';
import { extractPatchMetadata } from './snyk-file';
import { applyPatchToFile } from './patch';
import { findPhysicalModules } from './explore-node-modules';
import {
  VulnIdAndPackageName,
  VulnPatches,
  PatchedModule,
  ProtectResultType,
} from './types';
import { getAllPatches } from './fetch-patches';
import { sendAnalytics } from './analytics';

async function protect(projectFolderPath: string) {
  const snykFilePath = path.resolve(projectFolderPath, '.snyk');

  if (!fs.existsSync(snykFilePath)) {
    console.log('No .snyk file found');
    sendAnalytics({
      type: ProtectResultType.NO_SNYK_FILE,
    });
    return;
  }

  const snykFileContents = fs.readFileSync(snykFilePath, 'utf8');
  const snykFilePatchMetadata = extractPatchMetadata(snykFileContents);
  const vulnIdAndPackageNames: VulnIdAndPackageName[] = snykFilePatchMetadata;
  const targetPackageNames = [
    ...new Set(snykFilePatchMetadata.map((vpn) => vpn.packageName)), // get a list of unique package names by converting to Set and then back to array
  ];

  // find instances of the target packages by spelunking through the node_modules looking for modules with a target packageName
  const foundPhysicalPackages = findPhysicalModules(
    projectFolderPath,
    targetPackageNames,
  );

  // Map of package name to versions (for the target package names).
  // For each package name, we might have found multiple versions and we'll need to fetch patches for each version.
  // We will use this to lookup the versions of packages to get patches for.
  const packageNameToVersionsMap = new Map<string, string[]>();
  foundPhysicalPackages.forEach((p) => {
    if (packageNameToVersionsMap.has(p.packageName)) {
      const versions = packageNameToVersionsMap.get(p.packageName);
      if (!versions?.includes(p.packageVersion)) {
        versions?.push(p.packageVersion);
      }
    } else {
      packageNameToVersionsMap.set(p.packageName, [p.packageVersion]);
    }
  });

  const packageAtVersionsToPatches: Map<
    string,
    VulnPatches[]
  > = await getAllPatches(vulnIdAndPackageNames, packageNameToVersionsMap);

  if (packageAtVersionsToPatches.size === 0) {
    console.log('Nothing to patch');
    sendAnalytics({
      type: ProtectResultType.NOTHING_TO_PATCH,
    });
    return;
  }

  const patchedModules: PatchedModule[] = [];
  foundPhysicalPackages.forEach((fpp) => {
    const packageNameAtVersion = `${fpp.packageName}@${fpp.packageVersion}`;
    const vuldIdAndPatches = packageAtVersionsToPatches.get(
      packageNameAtVersion,
    );
    vuldIdAndPatches?.forEach((vp) => {
      vp.patches.forEach((patchDiffs) => {
        patchDiffs.patchDiffs.forEach((diff) => {
          applyPatchToFile(diff, fpp.path);
        });
      });
      patchedModules.push({
        vulnId: vp.vulnId,
        packageName: fpp.packageName,
        packageVersion: fpp.packageVersion,
      });
    });
  });

  console.log('Successfully applied Snyk patches');

  sendAnalytics({
    type: ProtectResultType.APPLIED_PATCHES,
    patchedModules,
  });
}

export default protect;
