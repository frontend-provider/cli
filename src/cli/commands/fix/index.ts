export = fix;

import * as Debug from 'debug';
import * as snykFix from '@snyk/fix';

import { MethodArgs } from '../../args';
import * as snyk from '../../../lib';
import { TestResult } from '../../../lib/snyk-test/legacy';

import { convertLegacyTestResultToFixEntities } from './convert-legacy-tests-results-to-fix-entities';
import { formatTestError } from '../test/format-test-error';
import { processCommandArgs } from '../process-command-args';
import { validateCredentials } from '../test/validate-credentials';
import { validateTestOptions } from '../test/validate-test-options';
import { setDefaultTestOptions } from '../test/set-default-test-options';
import { validateFixCommandIsSupported } from './validate-fix-command-is-supported';

const debug = Debug('snyk-fix');
const snykFixFeatureFlag = 'cliSnykFix';

async function fix(...args: MethodArgs): Promise<string> {
  const { options: rawOptions, paths } = await processCommandArgs(...args);
  const options = setDefaultTestOptions(rawOptions);
  await validateFixCommandIsSupported(options);
  validateTestOptions(options);
  validateCredentials(options);

  const results: snykFix.EntityToFix[] = [];
  results.push(...(await runSnykTestLegacy(options, paths)));

  // fix
  debug(
    `Organization has ${snykFixFeatureFlag} feature flag enabled for experimental Snyk fix functionality`,
  );
  const { fixSummary, meta } = await snykFix.fix(results);

  if (meta.fixed === 0) {
    throw new Error(fixSummary);
  }
  return fixSummary;
}

/* @deprecated
 * TODO: once project envelope is default all code below will be deleted
 * we should be calling test via new Ecosystems instead
 */
async function runSnykTestLegacy(
  options,
  paths,
): Promise<snykFix.EntityToFix[]> {
  const results: snykFix.EntityToFix[] = [];
  for (const path of paths) {
    // Create a copy of the options so a specific test can
    // modify them i.e. add `options.file` etc. We'll need
    // these options later.
    const snykTestOptions = {
      ...options,
      path,
      projectName: options['project-name'],
    };

    let testResults: TestResult | TestResult[];

    try {
      testResults = await snyk.test(path, snykTestOptions);
    } catch (error) {
      const testError = formatTestError(error);
      throw testError;
    }
    const resArray = Array.isArray(testResults) ? testResults : [testResults];
    const newRes = convertLegacyTestResultToFixEntities(resArray, path);
    results.push(...newRes);
  }
  return results;
}
