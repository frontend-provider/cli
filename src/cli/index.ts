#!/usr/bin/env node
import 'source-map-support/register';

// assert supported node runtime version
import * as runtime from './runtime';
// require analytics as soon as possible to start measuring execution time
import * as analytics from '../lib/analytics';
import * as alerts from '../lib/alerts';
import * as sln from '../lib/sln';
import {args as argsLib} from './args';
import {copy} from './copy';
import spinner = require('../lib/spinner');
import errors = require('../lib/error');
import ansiEscapes = require('ansi-escapes');
import {isPathToPackageFile} from '../lib/detect';
import {updateCheck} from '../lib/updater';

async function runCommand(args) {
  const result = await args.method(...args.options._);

  const res = analytics({
    args: args.options._,
    command: args.command,
  });

  if (result && !args.options.quiet) {
    if (args.options.copy) {
      copy(result);
      console.log('Result copied to clipboard');
    } else {
      console.log(result);
    }
  }

  return res;
}

async function handleError(args, error) {
  spinner.clearAll();
  let command = 'bad-command';

  if (error.code === 'VULNS') {
    // this isn't a bad command, so we won't record it as such
    command = args.command;
  } else if (!error.stack) { // log errors that are not error objects
    analytics.add('error', JSON.stringify(error));
    analytics.add('command', args.command);
  } else {
    // remove vulnerabilities from the errors
    // to keep the logs small
    if (error.stack && error.stack.vulnerabilities) {
      delete error.vulnerabilities;
    }
    if (error.message && error.message.vulnerabilities) {
      delete error.message.vulnerabilities;
    }
    analytics.add('error-message', error.message);
    analytics.add('error', error.stack);
    analytics.add('error-code', error.code);
    analytics.add('command', args.command);
  }

  const res = analytics({
    args: args.options._,
    command,
  });

  if (args.options.debug) {
    console.log(error.stack);
  } else {
    if (!args.options.quiet) {

      const result = errors.message(error);
      if (args.options.copy) {
        copy(result);
        console.log('Result copied to clipboard');
      } else {
        if (`${error.code}`.indexOf('AUTH_') === 0) {
          // remove the last few lines
          const erase = ansiEscapes.eraseLines(4);
          process.stdout.write(erase);
        }
        console.log(result);
      }
    }
  }

  return res;
}

function checkRuntime() {
  if (!runtime.isSupported(process.versions.node)) {
    console.error(`${process.versions.node} is an unsupported nodejs ` +
      `runtime! Supported runtime range is '${runtime.supportedRange}'`);
    console.error('Please upgrade your nodejs runtime version and try again.');
    process.exit(1);
  }
}

// Check if user specify package file name as part of path
// and throw error if so.
function checkPaths(args) {
  for (const path of args.options._) {
    if (typeof path === 'string' && isPathToPackageFile(path)) {
      throw new Error(`Not a recognised option did you mean --file=${path}. ` +
        'Check other options by running snyk --help');
    }
  }
}

async function main() {
  updateCheck();
  checkRuntime();

  const args = argsLib(process.argv);

  if (args.options.file && args.options.file.match(/\.sln$/)) {
    sln.updateArgs(args);
  }

  let res = null;
  let failed = false;

  try {
    checkPaths(args);
    res = await runCommand(args);
  } catch (error) {
    failed = true;
    res = await handleError(args, error);
  }

  if (!args.options.json) {
    console.log(alerts.displayAlerts());
  }

  if (!process.env.TAP && failed) {
    process.exit(1);
  }

  return res;
}

const cli = main().catch((e) => {
  console.log('super fail', e.stack);
  process.exit(1);
});

if (module.parent) {
  module.exports = cli;
}
