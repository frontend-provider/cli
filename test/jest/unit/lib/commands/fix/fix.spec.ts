import * as pathLib from 'path';
import * as fs from 'fs';
import * as snykFix from '@snyk/fix';

import cli = require('../../../../../../src/cli/commands');
import * as snyk from '../../../../../../src/lib';
import * as featureFlags from '../../../../../../src/lib/feature-flags';
import * as analytics from '../../../../../../src/lib/analytics';

import stripAnsi from 'strip-ansi';

const testTimeout = 100000;

const pipAppWorkspace = pathLib.join(
  __dirname,
  '../../../../../',
  'acceptance',
  'workspaces',
  'pip-app',
);

const npmWorkspace = pathLib.join(
  __dirname,
  '../../../../../',
  'acceptance',
  'workspaces',
  'no-vulns',
);

const pipRequirementsTxt = pathLib.join(pipAppWorkspace, 'requirements.txt');

const pipRequirementsCustomTxt = pathLib.join(
  __dirname,
  '../../../../../',
  'acceptance',
  'workspaces',
  'pip-app-custom',
  'base.txt',
);

const pipWithRemediation = JSON.parse(
  fs.readFileSync(
    pathLib.resolve(
      __dirname,
      '../../../../../',
      'fixtures',
      'snyk-fix',
      'test-result-pip-with-remediation.json',
    ),
    'utf8',
  ),
);

const pipNoIssues = JSON.parse(
  fs.readFileSync(
    pathLib.resolve(
      __dirname,
      '../../../../../',
      'fixtures',
      'snyk-fix',
      'test-result-pip-no-vulns.json',
    ),
    'utf8',
  ),
);

describe('snyk fix (functional tests)', () => {
  let origStdWrite;
  let snykFixSpy: jest.SpyInstance;
  let addAnalyticsSpy: jest.SpyInstance;

  beforeAll(async () => {
    origStdWrite = process.stdout.write;
    jest
      .spyOn(featureFlags, 'isFeatureFlagSupportedForOrg')
      .mockResolvedValue({ ok: true });
  });

  beforeEach(() => {
    snykFixSpy = jest.spyOn(snykFix, 'fix');
    addAnalyticsSpy = jest.spyOn(analytics, 'add');
  });

  afterEach(() => {
    snykFixSpy.mockClear();
    addAnalyticsSpy.mockClear();
  });

  afterAll(async () => {
    process.stdout.write = origStdWrite;
    jest.clearAllMocks();
  });
  it(
    'shows successful fixes Python requirements.txt project was fixed via --file',
    async () => {
      let stdoutMessages = '';
      process.stdout.write = (str) => {
        stdoutMessages += str;
        return true;
      };

      jest.spyOn(snyk, 'test').mockResolvedValue({
        ...pipWithRemediation,
        // pip plugin does not return targetFile, instead fix will fallback to displayTargetFile
        displayTargetFile: pipRequirementsTxt,
      });
      const res = await cli.fix('.', {
        file: pipRequirementsTxt,
        dryRun: true, // prevents write to disc
        quiet: true,
      });
      expect(snykFixSpy).toHaveBeenCalledTimes(1);
      expect(snykFixSpy.mock.calls[0][1]).toEqual({
        dryRun: true,
        quiet: true,
      });
      expect(addAnalyticsSpy).toHaveBeenCalledTimes(8);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFailedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixVulnerableProjects',
        1,
      );
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixableIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalIssues', 4);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixSummary',
        expect.any(String),
      );

      expect(stripAnsi(res)).toMatch('✔ Upgraded Jinja2 from 2.7.2 to 2.11.3');
      expect(stdoutMessages).toEqual('');
    },
    testTimeout,
  );
  it(
    'shows successful fixes Python requirements.txt project on stdout',
    async () => {
      let stdoutMessages = '';
      process.stdout.write = (str) => {
        stdoutMessages += str;
        return true;
      };
      jest.spyOn(snyk, 'test').mockResolvedValue({
        ...pipWithRemediation,
        // pip plugin does not return targetFile, instead fix will fallback to displayTargetFile
        displayTargetFile: pipRequirementsTxt,
      });
      const res = await cli.fix('.', {
        file: pipRequirementsTxt,
        dryRun: true, // prevents write to disc
      });
      expect(snykFixSpy).toHaveBeenCalledTimes(1);
      expect(snykFixSpy.mock.calls[0][1]).toEqual({
        dryRun: true,
      });
      expect(addAnalyticsSpy).toHaveBeenCalledTimes(8);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFailedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixVulnerableProjects',
        1,
      );
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixableIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalIssues', 4);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixSummary',
        expect.any(String),
      );

      expect(stripAnsi(res)).toMatch('✔ Upgraded Jinja2 from 2.7.2 to 2.11.3');
      expect(stripAnsi(stdoutMessages)).toMatch(
        '✔ Looking for supported Python items',
      );
    },
    testTimeout,
  );
  it(
    'shows successful fixes Python custom name base.txt project was fixed via --file',
    async () => {
      // read data from console.log
      let stdoutMessages = '';
      let stderrMessages = '';
      jest
        .spyOn(console, 'log')
        .mockImplementation((msg: string) => (stdoutMessages += msg));
      jest
        .spyOn(console, 'error')
        .mockImplementation((msg: string) => (stderrMessages += msg));

      jest.spyOn(snyk, 'test').mockResolvedValue({
        ...pipWithRemediation,
        // pip plugin does not return targetFile, instead fix will fallback to displayTargetFile
        displayTargetFile: pipRequirementsCustomTxt,
      });
      const res = await cli.fix('.', {
        file: pipRequirementsCustomTxt,
        packageManager: 'pip',
        dryRun: true, // prevents write to disc
        quiet: true,
      });
      expect(snykFixSpy).toHaveBeenCalledTimes(1);
      expect(snykFixSpy.mock.calls[0][1]).toEqual({
        dryRun: true,
        quiet: true,
      });

      expect(addAnalyticsSpy).toHaveBeenCalledTimes(8);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFailedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixVulnerableProjects',
        1,
      );
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixableIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalIssues', 4);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixSummary',
        expect.any(String),
      );

      expect(stripAnsi(res)).toMatch('✔ Upgraded Jinja2 from 2.7.2 to 2.11.3');
      expect(stdoutMessages).toEqual('');
      expect(stderrMessages).toEqual('');
    },
    testTimeout,
  );

  it(
    'snyk fix continues to fix when 1 path fails to test with `snyk fix path1 path2` (exit code 0)',
    async () => {
      // read data from console.log
      let stdoutMessages = '';
      let stderrMessages = '';
      jest
        .spyOn(console, 'log')
        .mockImplementation((msg: string) => (stdoutMessages += msg));
      jest
        .spyOn(console, 'error')
        .mockImplementation((msg: string) => (stderrMessages += msg));

      jest
        .spyOn(snyk, 'test')
        .mockRejectedValueOnce(new Error('Failed to get npm dependencies'));
      jest.spyOn(snyk, 'test').mockResolvedValue({
        ...pipWithRemediation,
        // pip plugin does not return targetFile, instead fix will fallback to displayTargetFile
        displayTargetFile: pipRequirementsTxt,
      });
      const res = await cli.fix(npmWorkspace, pipAppWorkspace, {
        dryRun: true, // prevents write to disc
        quiet: true,
      });
      expect(snykFixSpy.mock.calls[0][1]).toEqual({
        dryRun: true,
        quiet: true,
      });
      expect(addAnalyticsSpy).toHaveBeenCalledTimes(8);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFailedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixVulnerableProjects',
        1,
      );
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixableIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedIssues', 3);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalIssues', 4);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixSummary',
        expect.any(String),
      );

      expect(snykFixSpy).toHaveBeenCalledTimes(1);
      expect(stripAnsi(res)).toMatch('✔ Upgraded Jinja2 from 2.7.2 to 2.11.3');
      // only use ora to output
      expect(stdoutMessages).toEqual('');
      expect(stderrMessages).toEqual('');
    },
    testTimeout,
  );

  it(
    'snyk fails to fix when all paths fails to test with `snyk fix path1 path2` (non 0 error code)',
    async () => {
      // read data from console.log
      let stdoutMessages = '';
      let stderrMessages = '';
      jest
        .spyOn(console, 'log')
        .mockImplementation((msg: string) => (stdoutMessages += msg));
      jest
        .spyOn(console, 'error')
        .mockImplementation((msg: string) => (stderrMessages += msg));

      jest
        .spyOn(snyk, 'test')
        .mockRejectedValue(new Error('Failed to get dependencies'));

      let res;
      try {
        await cli.fix(npmWorkspace, pipAppWorkspace, {
          dryRun: true, // prevents write to disc
          quiet: true,
        });
      } catch (error) {
        res = error;
      }
      expect(addAnalyticsSpy).toHaveBeenCalledTimes(8);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFailedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixVulnerableProjects',
        0,
      );
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixableIssues', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedIssues', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalIssues', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixSummary',
        expect.any(String),
      );

      expect(snykFixSpy).toHaveBeenCalledTimes(1);
      expect(snykFixSpy.mock.calls[0][1]).toEqual({
        dryRun: true,
        quiet: true,
      });
      expect(stripAnsi(res.message)).toMatch('No successful fixes');
      expect(stdoutMessages).toEqual('');
      expect(stderrMessages).toEqual('');
    },
    testTimeout,
  );

  it(
    'snyk succeeds to fix when no vulns `snyk fix path1` (exit code 0)',
    async () => {
      // read data from console.log
      let stdoutMessages = '';
      let stderrMessages = '';
      jest
        .spyOn(console, 'log')
        .mockImplementation((msg: string) => (stdoutMessages += msg));
      jest
        .spyOn(console, 'error')
        .mockImplementation((msg: string) => (stderrMessages += msg));

      jest.spyOn(snyk, 'test').mockResolvedValue({
        ...pipNoIssues,
        // pip plugin does not return targetFile, instead fix will fallback to displayTargetFile
        displayTargetFile: pipRequirementsTxt,
      });
      const res = await cli.fix(pipAppWorkspace, {
        dryRun: true, // prevents write to disc
        quiet: true,
      });
      expect(snykFixSpy).toHaveBeenCalledTimes(1);
      expect(snykFixSpy.mock.calls[0][1]).toEqual({
        dryRun: true,
        quiet: true,
      });
      expect(addAnalyticsSpy).toHaveBeenCalledTimes(8);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFailedProjects', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalProjects', 1);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixVulnerableProjects',
        0,
      );
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixableIssues', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixFixedIssues', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith('snykFixTotalIssues', 0);
      expect(addAnalyticsSpy).toHaveBeenCalledWith(
        'snykFixSummary',
        expect.any(String),
      );

      expect(stripAnsi(res)).toMatch('✔ No vulnerable items to fix');
      expect(stdoutMessages).toEqual('');
      expect(stderrMessages).toEqual('');
    },
    testTimeout,
  );
});
