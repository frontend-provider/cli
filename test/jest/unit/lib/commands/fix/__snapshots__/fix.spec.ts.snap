// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snyk fix (functional tests) shows successful fixes Python requirements.txt project on stdout 1`] = `
"⠋ ⠋ ℹ Running \`snyk test\` for 
⠋ Looking for supported Python items✔ Looking for supported Python items
⠋ Processing 1 requirements.txt items.✔ Processing 1 requirements.txt items.
⠋ ✔ Done
"
`;
